### PR TITLE
feat: metrics

### PR DESCRIPTION
**Pull request by AlexsJones**
This pull request adds prometheus metrics into the connect service using a custom library ( go-http-metrics) 

( I decided not to go with custom gauges and counters but to use the http-metrics lib ) 

An example of the metrics we are able to leverage through http/https/grpc

```
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.005"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.01"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.025"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.05"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.1"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.25"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="0.5"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="2.5"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="5"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="10"} 3
http_request_duration_seconds_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="+Inf"} 3
http_request_duration_seconds_sum{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service=""} 0.0008348739999999999
http_request_duration_seconds_count{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service=""} 3
# HELP http_requests_inflight The number of inflight requests being handled at the same time.
# TYPE http_requests_inflight gauge
http_requests_inflight{handler="/schema.v1.Service/ResolveString",service=""} 0
# HELP http_response_size_bytes The size of the HTTP responses.
# TYPE http_response_size_bytes histogram
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="100"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1000"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="10000"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="100000"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1e+06"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1e+07"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1e+08"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="1e+09"} 3
http_response_size_bytes_bucket{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service="",le="+Inf"} 3
http_response_size_bytes_sum{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service=""} 162
http_response_size_bytes_count{code="200",handler="/schema.v1.Service/ResolveString",method="POST",service=""} 3
```

The advantage here is when this is scrapped we can perform some interesting queries 

```
sum(
    rate(http_request_duration_seconds_count[30s])
) by (handler)
```

### CODE REVIEW

**Review by beeme1mr — COMMENTED** [review]
Do you have a rough idea about what kind of overhead this will introduce? What do you think about providing the ability to either explicitly enable or disable it?

File: cmd/start.go [hunk 1]
```diff
@@ -29,7 +30,7 @@ func init() {
 	// allows environment variables to use _ instead of -
 	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_")) // sync-provider becomes SYNC_PROVIDER
 	viper.SetEnvPrefix("FLAGD")                            // port becomes FLAGD_PORT
-
+	flags.Int32P(metricsPortFlagName, "m", 8014, "Port to serve metrics on")
```
Comment (beeme1mr) [inline cmd/start.go:33]: Once this is merged, we should add this to the Prometheus wiki page on port allocations. 

https://github.com/prometheus/prometheus/wiki/Default-port-allocations

File: docs/configuration.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ Supported flags are as follows (result of running `./flagd start --help`):
   -e, --evaluator string          Set an evaluator e.g. json (default "json")
   -h, --help                      help for start
   -p, --port int32                Port to listen on (default 8013)
+  -m --metrics-port int32         Port to set up metrics listener on (default 8014)
```
Comment (beeme1mr) [inline docs/configuration.md:13]:
```suggestion
  -m, --metrics-port int32        Port to set up metrics listener on (default 8014)
```

File: pkg/service/connect_metrics.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	ID      string
+}
+
+type MetricsRecorder struct {
+	httpRequestDurHistogram   *prometheus.HistogramVec
+	httpResponseSizeHistogram *prometheus.HistogramVec
+	httpRequestsInflight      *prometheus.GaugeVec
+}
+
+func (r MetricsRecorder) ObserveHTTPRequestDuration(_ context.Context,
```
[the remark this answers is not in the document]
Comment (AlexsJones) [reply pkg/service/connect_metrics.go:58]: I took this from `go-http-metrics` and loosely modified it, so happy to remove the context in the next set of fixes

[the remark this answers is not in the document]
Comment (AlexsJones) [reply pkg/service/connect_metrics.go]: removed

**Review by james-milligan — APPROVED** [review]

**Pull request merged**